### PR TITLE
TESTS: Stabilize Renegotiation Test

### LINKED ISSUE #33772: testRenegotiation got stuck waiting for a handshake
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/7/console failed, rather helpfully saying

```
java.lang.Exception: Test abandoned because suite timeout was reached.
	at __randomizedtesting.SeedInfo.seed([A18ECABAE90405C4]:0)
```

On closer inspection, this was because the test thread was stuck here:

https://github.com/elastic/elasticsearch/blob/88c178dca6587c6f7f0f7b78951284553df0d8b3/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SimpleSecurityNetty4TransportTests.java#L227

```
  2>  - waiting on java.util.concurrent.CountDownLatch$Sync@34374363
  2>  at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
  2>  at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
  2>  at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
  2>  at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
  2>  at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
  2>  at org.elasticsearch.xpack.security.transport.netty4.SimpleSecurityNetty4TransportTests.testRenegotiation(SimpleSecurityNetty4TransportTests.java:227)
  2>  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  2>  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  2>  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  2>  at java.lang.reflect.Method.invoke(Method.java:498)
  2>  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
  2>  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
  2>  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
  2>  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
  2>  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  2>  at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
  2>  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  2>  at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
  2>  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  2>  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  2>  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  2>  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
  2>  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
  2>  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
  2>  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
  2>  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
  2>  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
  2>  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
  2>  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
  2>  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  2>  at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
  2>  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  2>  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  2>  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  2>  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  2>  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  2>  at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  2>  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
  2>  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
  2>  at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
  2>  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  2>  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
  2>  at java.lang.Thread.run(Thread.java:748)
```

I see messages that possibly point towards issues:

```
  1> Caused by: java.security.cert.CertificateException: No subject alternative names matching IP address 127.0.1.1 found
  1>  at sun.security.util.HostnameChecker.matchIP(HostnameChecker.java:168) ~[?:?]
  1>  at sun.security.util.HostnameChecker.match(HostnameChecker.java:94) ~[?:?]
  2>  - locked sun.nio.ch.EPollSelectorImpl@36f83869
  1>  at sun.security.ssl.X509TrustManagerImpl.checkIdentity(X509TrustManagerImpl.java:455) ~[?:?]
  1>  at sun.security.ssl.X509TrustManagerImpl.checkIdentity(X509TrustManagerImpl.java:436) ~[?:?]
  1>  at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:252) ~[?:?]
  1>  at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:136) ~[?:?]
  1>  at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1601) ~[?:?]
  1>  at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:216) ~[?:?]
  1>  at sun.security.ssl.Handshaker.processLoop(Handshaker.java:1052) ~[?:?]
  1>  at sun.security.ssl.Handshaker$1.run(Handshaker.java:992) ~[?:?]
  1>  at sun.security.ssl.Handshaker$1.run(Handshaker.java:989) ~[?:?]
  1>  at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_181]
  1>  at sun.security.ssl.Handshaker$DelegatedTask.run(Handshaker.java:1467) ~[?:?]
  1>  at io.netty.handler.ssl.SslHandler.runDelegatedTasks(SslHandler.java:1435) ~[netty-handler-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1343) ~[netty-handler-4.1.28.Final.jar:4.1.28.Final]
  2>  at sun.nio.ch.SelectorImpl.select(SelectorImpl.java:97)
  1>  at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1177) ~[netty-handler-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1221) ~[netty-handler-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:489) ~[netty-codec-4.1.28.Final.jar:4.1.28.Final]
  2>  at io.netty.channel.nio.NioEventLoop.select(NioEventLoop.java:755)
  2>  at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:410)
  2>  at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884)
  2>  at java.lang.Thread.run(Thread.java:748)
```

and

```
  1> [2018-09-16T22:14:20,068][WARN ][i.n.c.DefaultChannelPipeline] [[nioEventLoopGroup-4-1]] An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
  1> io.netty.handler.codec.DecoderException: javax.net.ssl.SSLException: Received fatal alert: certificate_unknown
  1>  at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:459) ~[netty-codec-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265) ~[netty-codec-4.1.28.Final.jar:4.1.28.Final]
  2>  at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884)
  1>  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434) [netty-transport-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965) [netty-transport-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) [netty-transport-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:646) [netty-transport-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:546) [netty-transport-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:500) [netty-transport-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:460) [netty-transport-4.1.28.Final.jar:4.1.28.Final]
  2>  at java.lang.Thread.run(Thread.java:748)
  1>  at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884) [netty-common-4.1.28.Final.jar:4.1.28.Final]
  2> "elasticsearch[TS_A][[transport_server_worker.default]][T#13]" ID=4735 RUNNABLE (in native code)
  1>  at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.28.Final.jar:4.1.28.Final]
  1>  at java.lang.Thread.run(Thread.java:748) [?:1.8.0_181]
  1> Caused by: javax.net.ssl.SSLException: Received fatal alert: certificate_unknown
  1>  at sun.security.ssl.Alerts.getSSLException(Alerts.java:208) ~[?:?]
  1>  at sun.security.ssl.SSLEngineImpl.fatal(SSLEngineImpl.java:1666) ~[?:?]
  1>  at sun.security.ssl.SSLEngineImpl.fatal(SSLEngineImpl.java:1634) ~[?:?]
  1>  at sun.security.ssl.SSLEngineImpl.recvAlert(SSLEngineImpl.java:1800) ~[?:?]
  1>  at sun.security.ssl.SSLEngineImpl.readRecord(SSLEngineImpl.java:1083) ~[?:?]
  1>  at sun.security.ssl.SSLEngineImpl.readNetRecord(SSLEngineImpl.java:907) ~[?:?]
  1>  at sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:781) ~[?:?]
  1>  at javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:624) ~[?:1.8.0_181]
  1>  at io.netty.handler.ssl.SslHandler$SslEngineType$3.unwrap(SslHandler.java:294) ~[netty-handler-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1275) ~[netty-handler-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1177) ~[netty-handler-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1221) ~[netty-handler-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:489) ~[netty-codec-4.1.28.Final.jar:4.1.28.Final]
  1>  at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:428) ~[netty-codec-4.1.28.Final.jar:4.1.28.Final]
  1>  ... 16 more
```

but it's hard to be certain. Here's all the logs: [consoleText.txt.gz](https://github.com/elastic/elasticsearch/files/2389001/consoleText.txt.gz)

@tbrooks8 could you take a look?

### PR DESCRIPTION
**Pull request by original-brownbear**
* Handshake listener should be added before initiating the handshake
to avoid it not running if handshake is triggered before it is added
* The second `startHandshake` is not synchronous and a read of only
50ms may fail to trigger it entirely (the failure can be reproduced reliably by setting the socket timeout to `1`)
=> fixed by retrying the read until the handshake finishes (a longer timeout would've worked too,
but retrying  seemed more stable)
* Closes #33772

### CODE REVIEW

**Review by Tim-Brooks — COMMENTED** [review]
As I mentioned, I do not understand how your fix actually loops multiple times.

You could register a request handler with the service and send a request after renegotiation has been triggered and wait for the response back. Then we can stop this wait for timeout thing. Its just that our serialization work is currently pretty tightly coupled to the transport and you would manually have to serialize the request to send over a random socket.

**Comment by tbrooks8** [comment]
Also you can look into, maybe the `read` is not really necessary? It is possible that you `startHandshake` and then you try to write the `ping`. And the flush only succeeds once the renegotiation has been completed and the `ping` has been written?

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SimpleSecurityNetty4TransportTests.java
Comment (original-brownbear) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SimpleSecurityNetty4TransportTests.java]: :man_facepalming: sorry ... that was from when I verified this in my local experiements... fixing sec

**Comment by original-brownbear** [comment]
@tbrooks8 I tried triggering the renegotiation by writing the ping, didn't work (never completed the handshake, not even if I turn off Nagle ...). 
As far as I remember you need to invoke a `read` to trigger anything but the initial handshake to go through.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SimpleSecurityNetty4TransportTests.java
Comment (original-brownbear) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SimpleSecurityNetty4TransportTests.java]: Removed :)

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SimpleSecurityNetty4TransportTests.java
Comment (original-brownbear) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SimpleSecurityNetty4TransportTests.java]: Made it rethrow `AssertionError` so it isn't caught but we still see it logged even if we get hung up on the latch :)

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SimpleSecurityNetty4TransportTests.java
Comment (tbrooks8) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SimpleSecurityNetty4TransportTests.java]: Let's reduce the timeout now. Maybe 5-10 ms since we are retrying. No reason to require this test to take 50 ms when it will normally take much less (now that we are retrying).

**Review by Tim-Brooks — APPROVED** [review]
Let's merge this this after my PR and you can migrate it to the abstract class so it also fixes it for the nio security transport?

**Comment by original-brownbear** [comment]
Jenkins test this

**Pull request merged**
